### PR TITLE
Static semantics for records

### PR DESCRIPTION
**Pull request by leafpetersen**
This adds a proposed type inference specification, and incorporates records into the inference, subtyping, normalization, and upper and lower bounds documents.

cc @munificent @eernstg @lrhn @stereotype441 @jakemac53 @natebosch @kallentu @scheglov @srawlins @johnniwinther @chloestefantsova

### CODE REVIEW

File: resources/type-system/upper-lower-bounds.md [hunk 1]
```diff
@@ -173,6 +173,20 @@ We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as foll
...
+
+- **UP**(`(S0, ... Sk, {T0 d0, ..., Tn dn})`,
+         `(S0', ... Sk', {T0' d0, ..., Tn' dn})`) =
+   `(Q0, ...,Qk, {R0, ..., Rn})` if:
+     - `Qi` is **UP**(`Si`, `Si'`)
+     - `Ri` is **UP**(`Ti`, `Ti'`)
+
+- **UP**(`(...)`, `(...)`) = `Record` otherwise
+- **UP**(`(...)`, `T2`) = **UP**(`Object`, `T2`)
+- **UP**(`T1`, `(...)`) = **UP**(`T1`, `Object`)
```
Comment (munificent) [inline resources/type-system/upper-lower-bounds.md:188]: I'm guessing there's no point in using "**UP**(`Record`, `T2`)" and "**UP**(`T1`, `Record`)" here because by the time we reach these clauses we know the other type isn't `Record` and there's no other possible types that `T1` or `T2` could be where the choice between `Record` and `Object` would make a difference? I'm thinking if maybe `T1` is `Record?`. But nullable types are already handled before we get here, right?

**Review by munificent — APPROVED** [review]
This looks great!

File: accepted/future-releases/records/records-feature-specification.md [hunk 2]
```diff
@@ -308,6 +313,11 @@ no "row polymorphism" or "width subtyping".*
 
 ### Upper and lower bounds
 
+Bounds computatinos for record types has been incorporated into the main
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md]: Fixed.

File: resources/type-system/upper-lower-bounds.md (hunk 1 above)
Comment (leafpetersen) [reply resources/type-system/upper-lower-bounds.md:188]: We know these aren't nullable, there are no other cases that could apply.  Technically it might be more robust against refactoring to use `Record` here (in case we change something in such a way that it would matter), but this is consistent with `Function`, and I probably would lean towards just keeping it as is.

File: resources/type-system/inference.md [hunk 3]
```diff
@@ -564,6 +570,13 @@ replaced with `Never`, and every covariant occurrence of `Xi` replaced with
   from any of the `Bi`:
   - The least closure of `S` with respect to `L` is `Never`
   - The greatest closure of `S` with respect to `L` is `Function`
+- if `S` is `(T0 x0, ...., Tn xn,  {Tn+1 xn+1, ..., Tm xm})`:
+  - The least closure of `S` with respect to `L` is `(U0 x0, ...., Un1 xn, {Un+1
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply resources/type-system/inference.md]: Four dots means "really f****ing long list of things."  :)

Fixed.

File: accepted/future-releases/records/records-feature-specification.md [hunk 4]
```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
...
-Type inference and promotion flows through records in much the same way it does
-for instances of generic classes (which are covariant in Dart just like record
-fields are) and collection literals.
+Note that implicit casts are not considered to be applied as part of inference,
+hence:
+
+```dart
+  // Static error, inferred type of the record is (dynamic, double)
+  (num, num) r = (3 as dynamic, 3.5);
+```
```
Comment (eernstg) [inline accepted/future-releases/records/records-feature-specification.md:384]: OK, that's a natural consequence of using bottom-up inference: The resulting type is `(dynamic, double)`, so it isn't `dynamic`, so there's no implicit cast. Yet ...

Do we actually want to maintain that a cast from dynamic must be explicit in this particular situation? That feels like a surprising deviation from the approach otherwise taken in Dart (that is: by default, dynamic expressions get a cast implicitly; by lint or by `--something-strict`, developers can choose to handle every occurrence of the type dynamic explicitly).

This also means that we cannot apply the perspective that record-to-record assignability is treated in a way which is essentially field-by-field (and that perspective is otherwise applicable quite often).

Line 360-361 actually says that the context type schema is propagated down into record literals (similarly to the propagation of context type schemas into any other composite, e.g., list literals), so we could say that inference transforms an expression `e` with static type `dynamic` into `e as T`, where `T` is obtained from the context type schema in the same way as in those other cases:

```dart
X f<X>(X x) => x;

class C<X> {
  final X x;
  C(this.x);
}

void main() {
  var d = 1 as dynamic;
  int i = d; // Implicit cast, OK.
  int j = f(d); // Implicit cast, OK.
  C<int> c = C(d); // Implicit cast, OK.
  List<int> list = [d]; // Implicit cast, OK.
  Set<int> set = {d}; // Implicit cast, OK.
  Map<int, int> map = {d: d}; // Implicit casts, OK.
  (int,) r = (d,); // Error or implicit cast?
}
```

I believe this approach would also work if we wish to change type inference to be bottom-up with respect to soundly co/contra-variant type variables. The situation (and the surprise) would be the same:

```dart
class C<out X, inout Y> {
  final X x;
  Y y;
  C(this.x, this.y);
}

void main() {
  var d = 1 as dynamic;
  C<int, int> c = C(d, d); // Error or implicit casts?
}
```

**Review by eernstg — APPROVED** [review]
LGTM, noting one point that could be debated (about casts from dynamic).

File: accepted/future-releases/records/records-feature-specification.md (hunk 4 above)
Comment (chloestefantsova) [reply accepted/future-releases/records/records-feature-specification.md:384]: Maybe another minor argument in defence of casting from `dynamic` is the potential spreading of records into argument lists. In the following example the spreading will not work, forcing the manual spreading. The manual spreading may appear especially surprising in this case because it will work without the explicit downcast from `dynamic`, making the reader questioning why the spreading was done manually in the first place.

```dart
foo(Funcion(int, String) f, (dynamic, String) a) {
  // This will not work.
  f(...a); // '...a' is a hypothetical syntax for spreading arguments.

  // This will work.
  f(a.$0, a.$1); // The downcast from 'dynamic' will be inserted here.
}
```

**Review by chloestefantsova — APPROVED** [review]

File: accepted/future-releases/records/records-feature-specification.md [hunk 5]
```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
...
+
+```dart
+  // No error, inferred type of the record is (int, double)
+  (num, num) r = (3, 3.5)..$0.isEven;
+```
 
-Type inference and promotion flows through records in much the same way it does
-for instances of generic classes (which are covariant in Dart just like record
-fields are) and collection literals.
+Note that implicit casts are not considered to be applied as part of inference,
```
Comment (lrhn) [inline accepted/future-releases/records/records-feature-specification.md:378]: That's news to me. (Although I'll admit I wouldn't be able to point to where they are specified at all. I just always thought that the implicit downcasts were inserted during inference, like the missing type parameters and missing `.call` member accesses.)

I guess it's an acceptable distinction, since we allow *assignability* from one to the other, and we presumably remember the static type and context type found during inference for the rest of the life of the program (the runtime-semantics sometimes refers to them).

Which means that the implicit downcast is part of *runtime* semantics, and does not affect static typing.

I'm guessing `call` member tear-off is the same,
but double-literal conversion is not, and neither is implicit instantiated tear-off (mainly for historical reasons, today it probably could be):
```dart
T id<T>(T value) => value;
int Function(int) f = id..printStaticType(); 
```
prints `(int) -> int`.

File: accepted/future-releases/records/records-feature-specification.md [hunk 6]
```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
...
+hence:
+
+```dart
+  // Static error, inferred type of the record is (dynamic, double)
+  (num, num) r = (3 as dynamic, 3.5);
```
Comment (lrhn) [inline accepted/future-releases/records/records-feature-specification.md:383]: I would not make this an error (as commented on in an issue).
I'd define a record type to be assignable to another record type of the same shape, if the field types are individually assignable.

File: resources/type-system/subtyping.md [hunk 7]
```diff
@@ -43,6 +44,8 @@ The set of types under consideration are as follows:
 - Function types
   - `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, [Tn+1 xn+1, ..., Tm xm])`
   - `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, {Tn+1 xn+1, ..., Tm xm})`
+- Record types
+  - `(T0 x0, ...., Tn xn, {Tn+1 xn+1, ..., Tm xm})`
```
Comment (lrhn) [inline resources/type-system/subtyping.md:48]: Can drop the `x0`..`xn` names here. Or maybe say that they are optional and have no effect on anything.

File: resources/type-system/upper-lower-bounds.md (hunk 1 above)
Comment (lrhn) [reply resources/type-system/upper-lower-bounds.md:188]: This way of writing it implicitly enforces that `Record` cannot have a subtype which is not a record type. There are no non-record types between `Record` and a record type, and no types between `Record` and `Object`.

File: accepted/future-releases/records/records-feature-specification.md (hunk 4 above)
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:384]: > OK, that's a natural consequence of using bottom-up inference: The resulting type is (dynamic, double), so it isn't dynamic, so there's no implicit cast. Yet ...

Github PR comment locations aren't precise enough for me to know what you are referring to by "that".

> Do we actually want to maintain that a cast from dynamic must be explicit in this particular situation? That feels like a surprising deviation from the approach otherwise taken in Dart (that is: by default, dynamic expressions get a cast implicitly; by lint or by --something-strict, developers can choose to handle every occurrence of the type dynamic explicitly).

The approach I have taken here is consistent with what we currently do elsewhere.  The alternatives are a deviation.  I expand on this [here](https://github.com/dart-lang/language/issues/2488), let's continue discussion there.

File: resources/type-system/subtyping.md [hunk 8]
```diff
@@ -241,6 +246,10 @@ promoted type variables `X0 & S0` and `T1` is `X0 & S1` then:
   - and `B0i[Z0/X0, ..., Zk/Xk] === B1i[Z0/Y0, ..., Zk/Yk]` for `i` in `0...k`
   - where the `Zi` are fresh type variables with bounds `B0i[Z0/X0, ..., Zk/Xk]`
 
+- **Record Types**: `T0` is `(V0 x0, ..., Vn xn, {Vn+1 dn+1, ..., Vm xm})`
+  - and `T1` is `(S0 y0, ..., Sn yn, {Sn+1 yn+1, ..., Sm ym})`
+  - and `Vi <: Si` for `i` in `0...m`
+
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply resources/type-system/subtyping.md:252]: Insufficiently fixed cut & paste.  Fixed now, thanks!

File: accepted/future-releases/records/records-feature-specification.md (hunk 6 above)
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:383]: Discussion moved to #2488

File: accepted/future-releases/records/records-feature-specification.md [hunk 9]
```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
...
+### Type inference
+
+As usual, we define type inference for record expressions with respect to a
+context type schema which is determined by the surrounding context of the
+inferred expression.  Given a type schema `K` and a record expression `E` of the
+general form `(e0, ..., en, d0 : t0, ..., dm : tm)`, then inference proceeds as
+follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : P0, ...., dm :
+Pm})`, then:
```
[the remark this answers is not in the document]
Comment (stereotype441) [reply accepted/future-releases/records/records-feature-specification.md:359]: Is it necessary for the implementations to perform this sorting when normalizing types?  If so, it's probably worth a short mention in non-normative text, just so that the implementation teams don't forget to do so.

File: accepted/future-releases/records/records-feature-specification.md (hunk 4 above)
Comment (eernstg) [reply accepted/future-releases/records/records-feature-specification.md:384]: Just for the record: "that" is the error in line 382.

File: resources/type-system/upper-lower-bounds.md [hunk 10]
```diff
@@ -280,6 +294,16 @@ follows.
 
 - **DOWN**(`T Function<...>(...)`, `S Function<...>(...)`) = `Never` otherwise
 
+- **DOWN**(`(...)`, `Record`) = `(...)`
+- **DOWN**(`Record`, `(...)`) = `(...)`
```
Comment (scheglov) [inline resources/type-system/upper-lower-bounds.md:298]: It looks that we don't have similar statements for function types, at least in the analyzer. And I don't see them here. Instead it works because we use `DOWN(T1, T2) = T1 if T1 <: T2` and `DOWN(T1, T2) = T2 if T2 <: T1`.

File: accepted/future-releases/records/records-feature-specification.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (scheglov) [reply accepted/future-releases/records/records-feature-specification.md:359]: FWIW, it was more convenient for the analyzer to always ensure that named fields are sorted when we construct `RecordTypeImpl`. And positional fields are separated from named.

[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:359]: > Is it necessary for the implementations to perform this sorting when normalizing types?

It is not necessary, ordering and position of named fields is always irrelevant, just as they are for named parameters for function types already.  I will add commentary to this effect, since it's causing confusion.

[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:374]: I've added commentary to this effect to all relevant files.

**Comment by leafpetersen** [comment]
I have updated this to reflect the outcome of the discussion in #2488, PTAL.

**Review by munificent — APPROVED** [review]

File: resources/type-system/upper-lower-bounds.md (hunk 10 above)
Comment (leafpetersen) [reply resources/type-system/upper-lower-bounds.md:304]: Yeah, that's cleaner.  Fixed.

File: accepted/future-releases/records/records-feature-specification.md [hunk 11]
```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
...
+
+Similarly, function and method invocations with named arguments and records with
+named field entries are written with their named entries in an unspecified
+canonical order and position.  Unless otherwise called out, position of named
+entries is semantically irrelevant, and all invocations and record literals with
+the same named entries (possibly in different orders or locations) and the same
+positional entries are considered equivalent.
+
+Given a type schema `K` and a record expression `E` of the general form `(e0,
+..., en, d0 : e{n+1}, ..., dm : e{m+n+1})` inference proceeds as follows.
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:377]: > (e1,..., en, d1: e{n+1}, ..., dm: e{n+m})
> for reading. Then n is the number of positional fields and m is the number of named fields.

Done. 

> (Also, if there are no named fields, is m = -1? 😉)

I know, but I'm trying to compromise between formality and legibility here.  I can write it up using hand rolled vector notation, which I think helps nobody (or more likely, add three additional clauses: n, m are zero, n is zero, m is zero) or I can just do this, which I think (hope?) everybody understands the intent of?  And communicating intent is my point here.  If anybody is sincerely confused, I'll add the base cases.

File: accepted/future-releases/records/records-feature-specification.md [hunk 12]
```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
...
+dm : K{m+n+1})` then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+    - Let `Ri` be the greatest closure of `Ki`
+    - If `Ti` is a subtype of `Ri` then let `Ti'` be `Ti`
+    - Otherwise, if `Ti` is `dynamic`, then we insert an implicit cast on `ei`
+      to `Ri`, and let `Ti'` be `Ri`
+    - Otherwise, if `Ti` is coercible to `Ri` (via int-to-double conversion,
+      call method tearoff, or implicit generic instantiation), then we insert
+      the appropriate implicit coercion on `ei` to coerce it to `Ri`, and let
+      `Ti'` be `Ri`
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:389]: > (We can't use "int-to-double conversion" here. That must have happened to ei during its own inference. It's too late if that inference already inferred int.)

I agree that there is some oddity here, in that we haven't specified where coercions are done elsewhere.  That said, I don't understand your point here.  Can you say more about why we can't do this coercion here?  I think it is likely not *necessary* (because in practice it will have already been done), but it's certainly possible.  Two illustrative examples:

```
// Here, coercion has to happen inside the cascade.  `3..runtimeType` has type double, 
// no coercion implied by record inference 
(double d,) = (3..runtimeType,); 
// Here, we could say that the coercion has happened as part of inference of `3` (in which case no coercion
// is implied via record inference), or we could say that the coercion happens as part of inference of the record.  
(double d,) = (3,);
```

So I think that for int to double coercions, including them here is redundant/unnecessary, but I don't think it's wrong,  and in the absence of an actual specification of where int to double conversions are done, it felt safer to include them here.  For the remaining coercions, I think we do need to specify that they are done here, since I believe that they are not currently done eagerly.

> We need to say "via one or more of" for the other conversions.

I do not believe that we perform multiple conversions in one step (and the beliefs of the CFE and analyzer seem to be in line with mine).  Am I missing something here?

> I also think we should not let Ti be Ri here. It should instead be the actual, precise type of the coerced expression.

Makes sense.  Done.

File: accepted/future-releases/records/records-feature-specification.md [hunk 13]
```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
...
+positional entries are considered equivalent.
+
+Given a type schema `K` and a record expression `E` of the general form `(e0,
+..., en, d0 : e{n+1}, ..., dm : e{m+n+1})` inference proceeds as follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : K{n+1}, ....,
+dm : K{m+n+1})` then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+    - Let `Ri` be the greatest closure of `Ki`
+    - If `Ti` is a subtype of `Ri` then let `Ti'` be `Ti`
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:383]: It doesn't say that, it says let `Ti'` be `Ti`.  Agreed it's easy to miss the `'` though.  In general, I run out of meta-variables pretty quickly in these rules if I try to use a new letter for each step, but here there's few enough that I can do so.  So done.

File: accepted/future-releases/records/records-feature-specification.md [hunk 14]
```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
...
+  // Inferred type of the record is (int, double, int Function(int))
+  var c = Callable();
+  dynamic d = 3;
+  (num, double, int Function(int), void Function()) r = (d, 3, id, c);
+```
+and the record initialization in the last line above is implicitly coerced to be
+the equivalent of:
+```dart
+  (num, double, int Function(int), void Function()) r =
+     (d as num, 3.0, id<int>, c.call);
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:423]: Ack.

File: accepted/future-releases/records/records-feature-specification.md [hunk 15]
```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
...
+named field entries are written with their named entries in an unspecified
+canonical order and position.  Unless otherwise called out, position of named
+entries is semantically irrelevant, and all invocations and record literals with
+the same named entries (possibly in different orders or locations) and the same
+positional entries are considered equivalent.
+
+Given a type schema `K` and a record expression `E` of the general form `(e0,
+..., en, d0 : e{n+1}, ..., dm : e{m+n+1})` inference proceeds as follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : K{n+1}, ....,
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md:379]: I think my concern here is that a fair bit of the time, we're working with the raw markdown and using the html notation renders it pretty unreadable.  I kind of feel that once we go to that level, we should just be using latex or html.  Would it help if I consistently used `_`?  That is `K_0, ... K_{n+1}`?

File: accepted/future-releases/records/records-feature-specification.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: I think all my confusion came from the 0..n indices. Doing 1..n solves that perfectly.

The rest should be understandable in context.

File: accepted/future-releases/records/records-feature-specification.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: > in the absence of an actual specification of where int to double conversions are done

Where that coercion happens is completely specified. The semantics of a decimal literal with a context type which `double` is assignable to, and `int` is not, is a `double` value with static type `double`. 

The "double literal coercion" is not a value coercion. We never create an integer value at runtime, then convert it to double. Instead we change the semantics of the decimal number literal to directly evaluate to a double value. That's why it cannot happen later than at the evaluation of literal expression itself.

The distinction is important on native, where `double x = 0x10000000000000000;` is valid with value pow(2.0, 64), but going through an integer would be an error &mdash; `int x = 0x10000000000000000;` is a compile-time error because the numerical value of the literal doesn't fit in an `int`.

And in the same way, `double d = 0x3333333333333333;` is also an error because the numerical values doesn't fit in a double (so it's not interpreted as a double, and you get an error that `int` is not assignable to `double`).

(On the web, all values are double anyway.)

It *would* be invalid to do a double conversion at this point, because there is no conversion which is valid at the point where you already have an `int`-type subexpression.

[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: > I do not believe that we perform multiple conversions in one step

This works today, no warnings in DartPad or analyzer, and runs in the VM.
```dart
class Id {
  T call<T>(T value) => value;
}
int Function(int) f = Id();
void main() {
  print(f.runtimeType); // (int) => int
}
```
We perform both `.this` tear-off and `<int>` instantiation on the same expression.

In practice, we should reuse whatever specification we have for other expressions inside record literals too, so we get the same behavior. Or say we want the same behavior.
(We'll get rid of `.call` tear-off anyway, so we'll probably be down to one coercion when records launch anyway.)

File: accepted/future-releases/records/records-feature-specification.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: That would probably remove any possible confusion. 

I think it's acceptable without it too, as long as we are aware that it's a little ambiguous.
We're not trying to defend against an adversarial reader here. That's for the formal specification to worry about.

File: accepted/future-releases/records/records-feature-specification.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/records/records-feature-specification.md]: I missed the `'` too. :-/

File: accepted/future-releases/records/records-feature-specification.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md]: > This works today, no warnings in DartPad or analyzer, and runs in the VM.

Interesting.  This changed - I was testing in the 2.9 language version, and it was rejected.  But you are correct, this works on bleeding edge.  I don't know how or why this changed, did we actually make this change deliberately?  I don't think this is a good thing in general.  It's probably ok now given the limited set of coercions that we have, but if we ever add more coercions, I think we would almost certainly want to avoid multi-step coercions - it opens up all sorts of failure modes (e.g. ordering issues).

[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md]: Ack.  And just in the interest of clarification, from my perspective, "Otherwise, if Ti is coercible to Ri (via int-to-double conversion" in no way circumscribes the conversion to only look at the type.  A thing is only coercible via int-to-double conversion if the expression to be coerced is a literal.  It would be better if I had said `ei` is coercible from `Ti` to `Ri` I guess?

[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: Pedantically: No.
Mentioning `Ti` is the problem. The moment you know the static type of `ei`, type inference for `ei` has happened, and therefore it must *already* have been coerced to a double literal. If it hasn't, it's too late now anyway.

The other conversions (`.call`, generic instantiation, implicit downcast) are based on the static type and context type of an expression, and happens after inferring the static type (and everything type inference otherwise entails) on the expression. At runtime, the expression itself evaluates normally, then we do something to its value afterwards (`.call`/`<...>`/`as Ri`).

The decimal-numeral-to-double interpretation is based on the actual syntax of the expression and the context type, and happens *before* the static type is inferred. Then the static type becomes `double`. At runtime, we evaluate the literal directly to a double value. (And/or at compile time, since it's effectively constant.)

That's why there is no static type which is coercible to anything via int-to-double-conversion.

You can perhaps say, "if `ei` is coercible to `Ri` via ....", but it's easier to just not mention the (imprecisely named) int-to-double-conversion along with the other coercions that we may apply to the value of `e1`. Just make sure to propagate context types down, then the double interpretation will automatically be performed at the literal, by the literal. It's not something to worry about during upwards propagation at all. 

Basically, it's not a coercion at all. It's just that the static and dynamic semantics of an integer numeral depends on the expression's context type. Then it's just the plain semantics of the expression.

[the remark this answers is not in the document]
Comment (leafpetersen) [reply accepted/future-releases/records/records-feature-specification.md]: > The moment you know the static type of ei, type inference for ei has happened, and therefore it must already have been coerced to a double literal. If it hasn't, it's too late now anyway.

Can you give a concrete example of what it means to be "too late" here?

[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/records/records-feature-specification.md]: I guess that I'm saying that if an expression with context type `double` has been inferred a static type of `int`, then that is it. There is no later coercion which allows that `int` expression to be valid in a `double` context.

I don't think there is any example where the static type is `int`, the context type is `double`, and the expression is an integer literal, so we'd have to look at another expression than a literal.
```dart
int a = 1;
double x = a;
````
The static type of `a` is `int`, even in a `double` context. No coercion applies to that static type or that integer value.
It's simply a compile-time error to have an `int` expression in a `double` context.

(A remarkable number of expressions actually *do* work, including `double x = 1 + 1;`, which puts the second `1` in a `double` context.)

Or, to be short: "int-to-double" is not a coercion which applies to the value of an expression. It's simply a name we gave to the feature which changed the semantics of some integer literal expressions when they occur in a `double`-accepting context. Just like the set-or-map literal `{}` evaluates to either a set or a map depending on the context type.

The actual coercions are something we do to the resulting value of an expression, which is otherwise evaluated normally, in order to make the value match the expected context type. We insert behavior, statically and at runtime, to *make* the types match.

With integer literals, we could rename them to `integer-or-double-literal`, like  `map-or-set-literal`, and then we probably wouldn't be thinking about it as a coercion at all.
(Or we can think of it as an inferred change to the program, from integer literal to double literal, like the added inferred type arguments to `{}` which depends on the context type, but are effectively always there after type inference.)
